rust/main/chains/hyperlane-ethereum/src/rpc_clients/fallback/tests.rs

```diff
@@ -0,0 +1,297 @@
+    serde_json::from_str("not-a-json").map_err(|e| HttpClientError::SerdeJson {
+        err: e,
+        text: "".to_owned(),
+    })
+}
+
+#[async_trait]
+impl JsonRpcClient for EthereumProviderMock {
+    type Error = HttpClientError;
+
+    /// Pushes the `(method, params)` to the back of the `requests` queue,
+    /// pops the responses from the back of the `responses` queue
+    async fn request<T: Debug + Serialize + Send + Sync, R: DeserializeOwned>(
+        &self,
+        method: &str,
+        params: T,
+    ) -> Result<R, Self::Error> {
+        self.push(method, params);
+        if let Some(sleep_duration) = self.provider.request_sleep() {
+            sleep(sleep_duration).await;
+        }
+        if self.block_number.is_none() {
+            dummy_error_return_value()
+        } else {
+            dummy_success_return_value(self.block_number.unwrap())
+        }
+    }
+}
+
+impl PrometheusJsonRpcClientConfigExt for EthereumProviderMock {
+    fn node_host(&self) -> &str {
+        todo!()
+    }
+
+    fn chain_name(&self) -> &str {
+        todo!()
+    }
+}
+
+impl<C> EthereumFallbackProvider<C, JsonRpcBlockGetter<C>>
+where
+    C: JsonRpcClient<Error = HttpClientError>
+        + PrometheusJsonRpcClientConfigExt
+        + Into<JsonRpcBlockGetter<C>>
+        + Clone,
+    JsonRpcBlockGetter<C>: BlockNumberGetter,
+{
+    async fn fallback_test_call(&self) -> u64 {
+        self.request::<_, u64>(BLOCK_NUMBER_RPC, ()).await.unwrap()
+    }
+}
+
+#[tokio::test]
+async fn test_multicast_first_provider_succeeds_immediately() {
+    let fallback_provider_builder = FallbackProviderBuilder::default();
+    let providers = vec![
+        EthereumProviderMock::new(None, Some(1)),
+        EthereumProviderMock::new(None, Some(2)),
+        EthereumProviderMock::new(None, Some(3)),
+    ];
+    let fallback_provider = fallback_provider_builder.add_providers(providers).build();
+    let ethereum_fallback_provider = EthereumFallbackProvider::new(fallback_provider);
+    let provider_id = ethereum_fallback_provider
+        .multicast::<_, u64>(BLOCK_NUMBER_RPC, ())
+        .await
+        .unwrap();
+    let provider_call_count: Vec<_> =
+        ProviderMock::get_call_counts(&ethereum_fallback_provider).await;
+    assert_eq!(provider_id, 1);
+    assert_eq!(provider_call_count, vec![1, 0, 0]);
+}
+
+#[tokio::test]
+async fn test_multicast_second_provider_succeeds_immediately() {
+    let fallback_provider_builder = FallbackProviderBuilder::default();
+    let providers = vec![
+        EthereumProviderMock::new(None, None),
+        EthereumProviderMock::new(None, Some(2)),
+        EthereumProviderMock::new(None, Some(3)),
+    ];
+    let fallback_provider = fallback_provider_builder.add_providers(providers).build();
+    let ethereum_fallback_provider = EthereumFallbackProvider::new(fallback_provider);
+    let provider_id = ethereum_fallback_provider
+        .multicast::<_, u64>(BLOCK_NUMBER_RPC, ())
```

Comment:
where we could still improve coverage is making sure `.request` only uses multicast for `METHOD_SEND_RAW_TRANSACTION`. IMO in each of these unit tests we could call `.request(METHOD_SEND_RAW_TRANSACTION)` instead of `.multicast` and return a similar value that's unique in each provider (like we're doing with the block number current, but it could be any arbitrary string really)

I like how the fallback provider unit tests use `fallback_test_call` to call a method that's realistically expected to be routed to the fallback logic (`BLOCK_NUMBER_RPC`)

Reply:
Good point. I added `multicast_test_call` and now we cover the decision on multicast/fallback choice.